plans/restore.pp

```diff
@@ -0,0 +1,238 @@
+
+  # Path to the recovery tarball
+  Pattern[/.*\.tar\.gz$/] $input_file,
+) {
+  peadm::assert_supported_bolt_version()
+
+  $recovery_opts = (peadm::recovery_opts_default() + $restore)
+  $cluster = run_task('peadm::get_peadm_config', $targets).first.value
+  $arch = peadm::assert_supported_architecture(
+    getvar('cluster.params.primary_host'),
+    getvar('cluster.params.replica_host'),
+    getvar('cluster.params.primary_postgresql_host'),
+    getvar('cluster.params.replica_postgresql_host'),
+    getvar('cluster.params.compiler_hosts'),
+  )
+
+  $primary_target   = peadm::get_targets(getvar('cluster.params.primary_host'), 1)
+  $replica_target   = peadm::get_targets(getvar('cluster.params.replica_host'), 1)
+  $compiler_targets = peadm::get_targets(getvar('cluster.params.compiler_hosts'))
+
+  # Determine the array of targets to which the PuppetDB PostgreSQL database
+  # should be restored to. This could be as simple as just the primary server,
+  # or it could be two separate PostgreSQL servers.
+  $puppetdb_postgresql_targets = peadm::flatten_compact([
+      getvar('cluster.params.primary_postgresql_host') ? {
+        undef   => $primary_target,
+        default => peadm::get_targets(getvar('cluster.params.primary_postgresql_host'), 1),
+      },
+      getvar('cluster.params.replica_postgresql_host') ? {
+        undef   => $replica_target,
+        default => peadm::get_targets(getvar('cluster.params.replica_postgresql_host'), 1),
+      },
+  ])
+
+  $puppetdb_targets = peadm::flatten_compact([
+      $primary_target,
+      $replica_target,
+      $compiler_targets,
+  ])
+
+  $recovery_directory = "${dirname($input_file)}/${basename($input_file, '.tar.gz')}"
+# lint:ignore:strict_indent
+  run_command(@("CMD"/L), $primary_target)
+    umask 0077 \
+      && cd ${shellquote(dirname($recovery_directory))} \
+      && tar -xzf ${shellquote($input_file)}
+    | CMD
+# lint:endignore
+  # Map of recovery option name to array of database hosts to restore the
+  # relevant .dump content to.
+  $restore_databases = {
+    'orchestrator' => [$primary_target],
+    'activity'     => [$primary_target],
+    'rbac'         => [$primary_target],
+    'puppetdb'     => $puppetdb_postgresql_targets,
+  }.filter |$key,$_| {
+    $recovery_opts[$key] == true
+  }
+
+  if getvar('recovery_opts.classifier') {
+    out::message('# Restoring classification')
+    run_task('peadm::backup_classification', $primary_target,
+      directory => $recovery_directory
+    )
+    out::message("# Backed up current classification to ${recovery_directory}/classification_backup.json")
+
+    run_task('peadm::transform_classification_groups', $primary_target,
+      source_directory  => "${recovery_directory}/classifier",
+      working_directory => $recovery_directory
+    )
+
+    run_task('peadm::restore_classification', $primary_target,
+      classification_file => "${recovery_directory}/classification_backup.json",
```

Comment:
Should `classification_file` be `${recovery_directory}/transformed_classification.json` instead of `${recovery_directory}/classification_backup.json`?

https://github.com/puppetlabs/puppetlabs-peadm/blob/693940177b80d796fb57f3abcb11fd9cb9bc3367/tasks/transform_classification_groups.py#L10-L11

`peadm::transform_classification_groups` creates `classification_backup.json` (current classifier output) and `transformed_classification.json` (non-PE infra groups from source merged with current classifier output) in `working_directory`; presumably the intent is to restore the transformed output?

Reply:
I've tested and verified that this suggestion is correct.

Reply:
@golflimaechoecho Fixed this in the latest PR update.